src/Strategy/Python/PyProjectGeneric/Types.hs

```diff
@@ -0,0 +1,333 @@
+module Strategy.Python.PyProjectGeneric.Types
+  ( PyProjectType (..)
+  , LockFileType (..)
+  , PyProjectGeneric (..)
+  , PyProjectMetadata (..)
+  , PyProjectBuildSystem (..)
+  , PyProjectPoetry (..)
+  , PyProjectPDM (..)
+  , PoetryDependency (..)
+  , PyProjectDetailedVersionDependency (..)
+  , PyProjectGitDependency (..)
+  , PyProjectPathDependency (..)
+  , PyProjectUrlDependency (..)
+  , detectProjectType
+  , projectTypePriority
+  , prioritizeProjectType
+  , isPEP621
+  , isPoetry
+  , isPDM
+  ) where
+
+import Data.Aeson (ToJSON(..), object, (.=))
+import Data.Aeson.Types qualified as Aeson
+import Data.List (sortOn)
+import Data.Map (Map)
+import Data.Text (Text)
+import Data.Maybe (isJust)
+import GHC.Generics (Generic)
+import Strategy.Python.Util (Req)
+import Toml qualified
+import Toml.Schema qualified
+
+-- | Enumeration of PyProject types
+data PyProjectType
+  = PoetryProject
+  | PDMProject
+  | PEP621Project
+  | UnknownProject
+  deriving (Eq, Ord, Show, Generic)
+
+instance ToJSON PyProjectType where
+  toJSON = Aeson.genericToJSON Aeson.defaultOptions
+
+-- | Types of lock files
+data LockFileType 
+  = PoetryLock 
+  | PDMLock 
+  | UVLock     -- For future support
```

Comment:
We should design types for what we know now. We can add `UVLock` when we actually need it.